src/read/lazy.rs

```diff
@@ -0,0 +1,108 @@
+pub(crate) use imp::*;
+
+#[cfg(not(feature = "std"))]
+mod imp {
+    use alloc::sync::Arc;
+    use core::sync::atomic::{AtomicPtr, Ordering};
+    use core::{mem, ptr};
+
+    #[derive(Debug, Default)]
+    pub(crate) struct LazyArc<T> {
+        // Only written once with a value obtained from `Arc<T>::into_raw`.
+        value: AtomicPtr<T>,
+    }
+
+    impl<T> Drop for LazyArc<T> {
+        fn drop(&mut self) {
+            let value_ptr = self.value.load(Ordering::Acquire);
+            if !value_ptr.is_null() {
+                // SAFETY: all writes to `self.value` are pointers obtained from `Arc::into_raw`.
+                unsafe {
+                    Arc::from_raw(value_ptr);
+                }
+            }
+        }
+    }
+
+    impl<T> LazyArc<T> {
+        pub(crate) fn get<E, F: FnOnce() -> Result<T, E>>(&self, f: F) -> Result<Arc<T>, E> {
+            // Clone an `Arc` given a pointer obtained from `Arc::into_raw`.
+            // SAFETY: `value_ptr` must be a valid pointer obtained from `Arc<T>::into_raw`.
+            unsafe fn clone_arc_ptr<T>(value_ptr: *const T) -> Arc<T> {
+                let value = Arc::from_raw(value_ptr);
+                let clone = Arc::clone(&value);
+                mem::forget(value);
+                clone
+            }
+
+            // Return the existing value if already computed.
+            let value_ptr = self.value.load(Ordering::Acquire);
+            if !value_ptr.is_null() {
+                // SAFETY: all writes to `self.value` are pointers obtained from `Arc::into_raw`.
+                return Ok(unsafe { clone_arc_ptr(value_ptr) });
+            }
+
+            // Race to compute and set the value.
+            let value = f().map(Arc::new)?;
+            let value_ptr = Arc::into_raw(value);
+            match self.value.compare_exchange(
+                ptr::null_mut(),
+                value_ptr as *mut T,
+                Ordering::Release,
+                Ordering::Relaxed,
```

Comment:
I think some kind of explanation of why these orderings are chosen is called for here.

When I see `SeqCst` I feel okay (as far as I feel okay about atomics code in general!) and when I see `AcqRel` I know to double check that basically we are dealing with one memory location and not relying on orderings between operations on different memory locations, and when I see anything else I basically know that I am out of my depth. Sometimes I feel better if there is a link to some comment in `std`'s implementation that explains why weaker orderings are okay in this specific case, and how the code I am actually reading just does the same thing `std` is doing. But yeah I definitely don't have the hubris required to say I understand justifications for weaker orderings. Mostly I just try to make sure that whoever wrote the code actually really understands what they're doing and has a good justification, and the way I do that is by asking them to write a comment :)

Reply:
Good suggestion. I've improved my understanding and decided that `Relaxed` is wrong here, and I've added comments.